### PR TITLE
Add support for embedded palettes and other sprite metadata

### PR DESCRIPTION
**Pull request by pchote**
Games commonly stored useful metadata (e.g. palettes, weapon offsets, tileset data) alongside the pixel data in their sprite files.  We used to require modders to load this extra data using a custom trait, adding both complexity and load-time overheads.

This PR extends `ISpriteLoader` with support for returning an arbitrary metadata `TypeDictionary`, which can be exported via the sequence parser (and, in the future, the tilesert parser) for use by traits.

This plumbing is used to expose palettes embedded in png and r8 files via the `EmbeddedPalette` sequence entry and `PaletteFromEmbeddedSpritePalette` trait.

### CODE REVIEW

File: OpenRA.Mods.D2k/SpriteLoaders/R8Loader.cs [hunk 1]
```diff
@@ -57,9 +62,20 @@ public R8Frame(Stream s)
 
 				Data = s.ReadBytes(width * height);
 
-				// Ignore palette
+				// Palette palette
```
Comment (IceReaper) [inline OpenRA.Mods.D2k/SpriteLoaders/R8Loader.cs]: Should be ```// Read palette```

File: OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs [hunk 2]
```diff
@@ -93,6 +113,8 @@ public class DefaultSpriteSequence : ISpriteSequence
 		public int[] Frames { get; private set; }
 		public Rectangle Bounds { get; private set; }
 
+		public readonly uint[] EmbeddedPalette = null;
```
Comment (IceReaper) [inline OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs]: Initializing field by default value is redundant.

**Comment by pchote** [comment]
Depends on #15652.

File: OpenRA.Mods.Common/Traits/World/PaletteFromEmbeddedSpritePalette.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		public readonly string Image = null;
+
+		[FieldLoader.Require]
+		[Desc("Sequence holding the palette definition")]
+		public readonly string Sequence = null;
+
+		[Desc("Allow palette modifiers to change the palette.")]
+		public readonly bool AllowModifiers = true;
+
+		public object Create(ActorInitializer init) { return new PaletteFromEmbeddedSpritePalette(this); }
```
[the remark this answers is not in the document]
Comment (pchote) [reply OpenRA.Mods.Common/Traits/World/PaletteFromEmbeddedSpritePalette.cs]: Why do you want to subclass this trait? It does almost nothing!

**Comment by pchote** [comment]
Rebased and updated.  I don't have a testcase for embedded cursor palettes, so will need somebody to test this.

**Comment by reaperrr** [comment]
> PaletteFromEmbeddedSpritePaletteInfo must explicitly implement the interface member IProvidesCursorPaletteInfo.ReadPalette

**Comment by pchote** [comment]
Fixed.

**Review by abcdefg30 — DISMISSED** [review]

**Review by IceReaper — DISMISSED** [review]

File: OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs [hunk 4]
```diff
@@ -66,6 +69,11 @@ public bool TryParseSprite(Stream s, out ISpriteFrame[] frames)
 					Array.Copy(png.Data, frameStart + y * png.Width, frames[i].Data, y * frames[i].Size.Width, frames[i].Size.Width);
 			}
 
+			metadata = new TypeDictionary
+			{
+				new EmbeddedSpritePalette(png.Palette.Select(x => (uint)x.ToArgb()).ToArray())
```
[the remark this answers is not in the document]
Comment (pchote) [reply OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs]: Done.

**Review by IceReaper — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ea0133dd565686e1a16a45f1b48cb7c2a6bc6415

**Pull request merged**